File: ardoc/ardoc_cachescan.pl
#!/usr/bin/env perl
#
# ARDOC - NIghtly COntrol System
# Author Alex Undrus <undrus@bnl.gov>
# 
# ----------------------------------------------------------
# ardoc_cachescan.pl 
# ----------------------------------------------------------
use Env;
use locale;

if ($#ARGV < 1) {
warn "ardoc_cachescan.pl: error: at least two arguments required\n";  
exit 1;
}

my $ARDOC_HOME="$ARDOC_HOME";
my $ARDOC_WORK_AREA="$ARDOC_WORK_AREA";

$opt="";
$section="";
$output="${ARDOC_WORK_AREA}/cache_scan_result";
@conf_array=();

while ($#ARGV>=0) {
    $_ = $ARGV[0];
    /^-[cd]+$/ && /d/ && do { $opt="def"; $section=$ARGV[1]; shift; };
    /^-[cd]+$/ && /c/ && do { $opt="com"; $section=$ARGV[1]; shift; };
    /^-[o]+$/ && /o/ && do { $output=$ARGV[1]; shift; };
    /^-[f]+$/ && /f/ && do { push(@conf_array,$ARGV[1]); shift; };
    shift;
}

if ( $opt eq "" || $section eq "" ){
    warn "ardoc_cachescan.pl: error: options -c <section> or -d <section> required\n";
    exit 1;
}

if ($#conf_array < 0){
    warn "ardoc_cachescan.pl: error: ardoc_cache files are not provided\n";
    exit 1; 
}

($sect = $section) =~ s/\s+/ /g;

system("touch $output");
open(TF, ">$output");

foreach $file (@conf_array){
@lines=();
if ( -f $file ){
  open(FR, "$file"); 
  chomp(@lines=<FR>);
  close(FR)
}

$label=0;
$label_bypass=0;
foreach $line (@lines){
    $line =~ s/\s+/ /g;
    $line =~ s/^\s+//g;
    $line =~ s/\s+$//g;
    $line_low = lc($line);
    $line_low =~ s/ //g;
    if ( $opt eq "com" ){
	$stt=1; $sbp=1; $sbpend=1;
        if ( $line_low =~ /^</ && $line_low !~ /^<bypass/ && $line_low !~ /^<\/bypass/ ) {$stt=0;}
        if ( $line_low =~ /^<bypass/ ){$sbp=0;} 
	if ( $line_low =~ /^<\/bypass/ ){$sbpend=0;}
        if ( $label == 1 && $stt == 0 ){
	  if ( $label_bypass == 1 ){
	    $label_bypass=0;
            print TF "fi\n";
	  }
       last;
        } # if ( $label == 1 && $stt == 0 ){
        if ( $label == 1 ){
#	    print "LLLL $line\n";
        if ( $line_low !~ /source/ && $line !~ /cmt/ && $line_low !~ /echo/ && $line_low !~ /broadcast/ && $line_low !~ /^if/ && $line_low !~ /^:/ && $line_low !~ /^make/ && $line_low !~ /fi[^a-zA-Z0-9]/ && $line_low !~ /stat=/ && $line_low !~ /^#/ && $line_low =~ /=/ ){
	     print TF "export ${line}\n";
        }
        elsif ( $line_low =~ /^cmtprojectpath=/ || $line_low =~ /^cmtroot=/ || $line_low =~ /^cmtpath=/ ){
	    print TF "export ${line}\n";
        }
        elsif ( $sbp == 0 ){ # if ( $line
          $label_bypass=1;
          print TF "if [ \"\${ARDOC_BYPASS}\" != \"yes\" ]; then :;\n"; 
        }
        elsif ( $sbpend == 0 ){ # if ( $line
	    $label_bypass=0;
            print TF "fi\n";
        } 
        else { # if ( $line
            if ( $line_low ne "" ){
            print TF "${line}\n";
	    }
	} # if ( $line
        } # if ( $label == 1       
	if ( $line =~ /^<${sect}/ ) { $label=1;}
    }
    if ( $opt eq "def" ){
      if ( $line =~ /^<${sect}/ ) {
      ($lin = $line) =~ s/[<>]//gs;
      ($li = $lin) =~ s/${sect}//gs;
      @varr = split(" ", $li);
	for $var (@varr){
#            print "VPPSST $var\n";
	    print TF "$var\n";
	}
      }
    }
}
}
close(TF);  

File: ardoc/ardoc_cachescan.py
#!/usr/bin/env python3
import os
import sys
import argparse
import re

def main():
    parser = argparse.ArgumentParser(description='ARDOC Cache Scanner')
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument('-d', '--def_section', metavar='SECTION', help='Define section')
    group.add_argument('-c', '--com_section', metavar='SECTION', help='Command section')
    parser.add_argument('-o', '--output', help='Output file')
    parser.add_argument('-f', '--files', nargs='+', required=True, help='Input cache files')

    args = parser.parse_args()

    opt = "def" if args.def_section else "com"
    section = args.def_section if args.def_section else args.com_section
    
    output_file = args.output if args.output else os.path.join(os.environ.get("ARDOC_WORK_AREA", "."), "cache_scan_result")

    if not section:
        print("ardoc_cachescan.py: error: a section must be provided with -c or -d", file=sys.stderr)
        sys.exit(1)

    sect = re.sub(r'\s+', ' ', section)

    with open(output_file, 'w') as tf:
        for file_path in args.files:
            if not os.path.isfile(file_path):
                continue

            with open(file_path, 'r') as fr:
                lines = fr.readlines()

            in_section = False
            in_bypass = False
            for line in lines:
                line = line.strip()
                line_low = line.lower()
                
                if opt == "com":
                    # Check for section boundaries
                    if line_low.startswith('<') and not line_low.startswith('</bypass') and not line_low.startswith('<bypass'):
                        if in_section:
                            if in_bypass:
                                tf.write("fi\n")
                                in_bypass = False
                            break # End of the current section
                    
                    if in_section:
                        if line_low.startswith('<bypass'):
                            in_bypass = True
                            tf.write('if [ "${ARDOC_BYPASS}" != "yes" ]; then :;\n')
                            continue
                        if line_low.startswith('</bypass'):
                            if in_bypass:
                                tf.write("fi\n")
                            in_bypass = False
                            continue

                        # Process line content
                        if (re.search(r'=', line_low) and 
                            'source' not in line_low and 
                            'cmt' not in line and 
                            'echo' not in line_low and 
                            'broadcast' not in line_low and 
                            not line_low.startswith('if') and 
                            not line_low.startswith(':') and 
                            not line_low.startswith('make') and 
                            not re.match(r'fi[^a-zA-Z0-9]', line_low) and 
                            not re.match(r'stat=', line_low) and 
                            not line_low.startswith('#')):
                            tf.write(f"export {line}\n")
                        elif (line_low.startswith('cmtprojectpath=') or 
                              line_low.startswith('cmtroot=') or 
                              line_low.startswith('cmtpath=')):
                            tf.write(f"export {line}\n")
                        else:
                            if line: # non-empty line
                                tf.write(f"{line}\n")

                    if line.startswith(f"<{sect}"):
                        in_section = True

                elif opt == "def":
                    if line.startswith(f"<{sect}"):
                        line_content = re.sub(r'[<>]', '', line)
                        line_content = re.sub(r'^\s*' + re.escape(sect) + r'\s*', '', line_content, flags=re.IGNORECASE)
                        variables = line_content.split()
                        for var in variables:
                            tf.write(f"{var}\n")

if __name__ == "__main__":
    main()
